qiskit/algorithms/minimum_eigen_solvers/qaoa/qaoa_circuit.py

```diff
@@ -0,0 +1,209 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Global X phases and parameterized problem hamiltonian."""
+
+from typing import Optional, Union, cast, Set
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import Parameter, ParameterVector
+from qiskit.circuit.library import BlueprintCircuit
+from qiskit.opflow import OperatorBase, I, X, CircuitStateFn, H, CircuitOp, EvolutionFactory
+
+
+class QAOACircuit(BlueprintCircuit):
+    """Global X phases and parameterized problem hamiltonian."""
+
+    def __init__(self,
+                 cost_operator: OperatorBase,
+                 param_p: int,
+                 initial_state: Optional[QuantumCircuit] = None,
+                 mixer_operator: Optional[Union[QuantumCircuit, OperatorBase]] = None,
+                 name: str = "qaoa"):
+        """
+        Constructor, following the QAOA paper https://arxiv.org/abs/1411.4028
+
+        Args:
+            cost_operator: The operator representing the cost of the optimization problem,
+                denoted as U(B, gamma) in the original paper.
+            param_p: The integer parameter p, which determines the depth of the circuit,
+                as specified in the original paper.
+            initial_state: An optional initial state to use.
+            mixer_operator: An optional custom mixer to use instead of the global X-rotations,
+                denoted as U(B, beta) in the original paper. Can be an operator or
+                an optionally parameterized quantum circuit.
+            name: A name of the circuit, default 'qaoa'
+        Raises:
+            TypeError: invalid input
+        """
+        super().__init__(name=name)
+        self._cost_operator = cost_operator
+        self._param_p = param_p
+        self._initial_state = initial_state
+        self._mixer_operator = mixer_operator
+
+        # set this circuit as a not-built circuit
+        self._num_qubits = 0
```

Comment:
This is something that needs to be done that is not say done by Blueprint circuit (I did not see it there - maybe this is more for @Cryoris around having to know/set this as a subclass - especially since it seems data being None seems to be more the indicator of a circuit not having been built)

Reply:
I think this is usually set to `None`

Reply:
Now `None`.